### ISSUE TITLE
Benchmark attention-sink pinning (and transformers-community/sink_cache) as quality/efficiency options for rolling generation

### ISSUE THREAD

**Issue by mmcdermott** [mmcdermott/MEDS_EIC_AR#90]
## Summary

Follow-up to #82 / #86: experiment with **attention-sink token pinning** as a quality / efficiency knob for the sliding-window rolling generation loop added in #86. This is **not a correctness fix** — the current loop's RoPE semantics are already in-distribution (see [the corrected framing](https://github.com/mmcdermott/MEDS_EIC_AR/issues/82#issuecomment-4253727889)) — but attention-sink pinning can plausibly reduce compounding-error artifacts near sliding boundaries on real pretrained checkpoints, and a drop-in backend like `transformers-community/sink_cache` can separately avoid the per-chunk re-prefill cost.

## Motivation

The current `_rolling_generate` loop is **cache-less**: every chunk discards the KV cache and re-calls `HF_model.generate(prompt, ...)` on the right-aligned tail of `(input ++ generated_so_far)`. This is mathematically clean — HF re-rotates surviving tokens at fresh RoPE positions that sit squarely in the model's trained range, so each chunk is a valid in-distribution prefix — but it has two residual concerns:

1. **Generic finite-context information loss.** Once the window slides, evicted tokens are permanently unavailable. Same as any fixed-window model; not specific to sliding.
2. **Compounding-error manifestations under shifted prefixes.** The trained model has learned patterns where generated tokens are conditioned on "whatever was at the beginning of the sequence." When the window slides and the start-of-sequence content changes, those patterns can produce subtly off output near chunk boundaries. Empirical, not mathematical.

Concern 2 is what StreamingLLM ([arXiv:2309.17453](https://arxiv.org/abs/2309.17453)) addresses with **attention sinks** — pinning a small number of early tokens in every chunk's visible prefix so the model always conditions on the same "start" content.

### Experiment 1: prompt-level attention-sink pinning (quality)

**Cheap version; no library dependency.** Add an `attention_sink_tokens: int = 0` kwarg to `_rolling_generate`. When `> 0`, every chunk's prompt becomes `input_ids[:, :attention_sink_tokens]` prepended to the sliding tail. So the pinned tokens always appear at positions `0 .. K-1` of each chunk's prompt, regardless of how many tokens we've generated.

**Pseudocode (within the existing loop):**

```python
K = attention_sink_tokens  # default 0 → current behavior
ctx_size = rolling_context_size if rolling_context_size is not None else max_seq_len - 1
tail_ctx = ctx_size - K  # remaining budget for the sliding tail

while n_generated < max_new_tokens and not bool(finished.all()):
    if K > 0:
        sink = input_ids[:, :K]
    # Build sliding tail as before, but capped at tail_ctx instead of ctx_size
    tail = _build_sliding_tail(input_ids, generated, n_generated, tail_ctx)
    prompt = torch.cat([sink, tail], dim=1) if K > 0 else tail
    ...
```

**Validation plan:**

- Run on a real pretrained MEDS-EIC-AR checkpoint (not the random-init doctest model — compounding error can't be measured on random init).
- Generate N trajectories per prediction point with three settings:
  - `attention_sink_tokens=0` (current behavior)
  - `attention_sink_tokens=4` (StreamingLLM default)
  - `attention_sink_tokens=8`
- Measure per setting:
  1. **Held-out likelihood on ground-truth continuations** where available.
  2. **Downstream task metrics from `meds-trajectory-evaluation`** on a task where longer-horizon generation is needed (so the rolling path is actually exercised).
  3. **Qualitative sanity check:** do the generated trajectories near sliding boundaries look reasonable, or is there visible degradation?
- Decide whether to make `attention_sink_tokens > 0` the default for rolling generation.

**Blast radius:** ~10 lines in `_rolling_generate`, plus one config field. No new dependencies. Safe to ship with a default of 0 (no behavior change) and let it be opt-in until validated.

### Experiment 2: cached-sliding backend via `transformers-community/sink_cache` (efficiency)

**The expensive version.** Our current loop pays for a full re-prefill on every chunk. If the model has `max_seq_len = S` and we're generating `T` total tokens via rolling, that's O(T/chunk_size · S²) attention operations for a single forward pass — an order of magnitude more compute than a KV-cache-reusing backend would need.

`transformers-community/sink_cache` is a community port of the pre-4.53.0 HF `SinkCache` class. It stores pre-rotation K/V in the cache, re-applies RoPE rotations at cache-read time with **re-assigned position IDs**, and supports incremental cached generation with a rolling eviction policy. Concretely:

- The first K tokens are pinned with fixed positions `0..K-1`.
- The remaining window slides: as new tokens are generated, the oldest non-sink cache entries are evicted.
- Surviving non-sink tokens are re-rotated at read time to reflect their *current* window slot, not their original emission position.
- The result is attention math that's equivalent to Experiment 1's prompt-level pinning, but achieved via cache reuse instead of re-prefill.

**Validation plan:**

- Verify `transformers-community/sink_cache` works with GPTNeoX (not guaranteed — the community port has to be tested).
- Integrate it behind a `cache_strategy` config flag in `_rolling_generate`. The outer rolling policy (finished mask, EOS stopping, time-budget stopping from #82) stays in Python; only the inner `HF_model.generate` call is swapped for a `past_key_values=sink_cache` variant.
- Benchmark: wall-clock for generating T tokens with the sink_cache backend vs. current cache-less backend. Expect a meaningful speedup (the prefill cost is the dominant factor in the current loop).
- Verify the outputs match or closely match Experiment 1's prompt-pinned outputs for the same `attention_sink_tokens` setting — they should be semantically equivalent.

**Blast radius:** larger. Requires: (1) confirming GPTNeoX compatibility, (2) integrating the cache object into the HF `generate` call site, (3) managing per-sample cache state across chunks in a batched loop, (4) handling the interaction with the `finished` mask for rows that EOS'd mid-batch (their cache state has to be frozen while other rows continue advancing). Non-trivial; worth doing only if Experiment 1 shows meaningful quality improvement that justifies the efficiency investment.

## Dependencies

- Experiment 1 can start today — no library dependencies, small diff, validated on any real pretrained checkpoint.
- Experiment 2 depends on (a) Experiment 1 showing it's worth chasing, and (b) `transformers-community/sink_cache` being compatible with GPTNeoX.
- Neither is a prerequisite for #82's time-budget stopping criterion, which lives in the outer Python loop and is backend-agnostic.

## Out of scope

- Swapping to a sliding-window-attention architecture upstream (Mistral/Gemma2/Llama4). That's a retraining project.
- Integrating vLLM or SGLang as the rolling backend — tracked separately in #87 / #88.

**Comment by mmcdermott** [comment]
Important clarification from a collaborator who built ETHOS-ARES:

> "The conflict between KV-caching and rolling windows is more severe than the current issue describes. Instead of the cache going from [a,b,c,d] to [b,c,d,e], the reality is that b in the first example will have a different embedding than b in the second window due to positional embeddings. This makes all KV-caching irrelevant and any attempts to 'slide' the KV-cache invalid. My general workaround for this in ETHOS-ARES was just to slide by some large number of tokens K so that you only have to refill the cache every K new tokens."

This confirms what we discovered during the PR #86 review: our current cache-less implementation (throw away KV cache each chunk, re-prefill from scratch) is the correct approach for GPTNeoX with RoPE. Attempting to "slide" the KV cache naively is invalid because positional embeddings change the cached values.

**The practical optimization the collaborator used — sliding by K tokens instead of 1** — is already supported by our `rolling_context_size` parameter. Setting `rolling_context_size` to something smaller than `max_seq_len - 1` means each chunk emits up to `max_seq_len - rolling_context_size` new tokens before the next re-prefill, amortizing the prefill cost over K tokens instead of 1.

SinkCache (Experiment 2 in this issue) would theoretically solve the positional-embedding problem by re-rotating at cache-read time, but the collaborator's experience suggests the large-stride approach is the practical workaround that actually works. I'd deprioritize the SinkCache experiment and instead document the `rolling_context_size` tuning knob as the recommended optimization path.

Reducing priority of this issue — it's a nice-to-have optimization, not a blocker for anything else.
